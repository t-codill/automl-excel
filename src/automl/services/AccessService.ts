import { AuthorizationManagementClient, AuthorizationManagementModels } from "@azure/arm-authorization";
import { includes, map, some } from "lodash";
import { IServiceBaseProps } from "./ServiceBase";
import { ServiceBaseArm } from "./ServiceBaseArm";

function hasPermission(permissionActions: Array<string[] | undefined>): boolean {
    return some(permissionActions, (permissionAction) => includes(permissionAction, "*"));
}

export class AccessService extends ServiceBaseArm<AuthorizationManagementClient> {
    constructor(props: IServiceBaseProps) {
        super(props, AuthorizationManagementClient);
    }

    public async getAccess(): Promise<AuthorizationManagementModels.PermissionGetResult | undefined> {
        return this.send(async (client, abortSignal) => {
            return client.permissions.listForResource(
                this.props.resourceGroupName,
                "Microsoft.MachineLearningServices",
                "",
                "workspaces",
                this.props.workspaceName, {
                    abortSignal
                });
        });
    }
    public readonly checkPermission = async () => {
        const curPermissions = await this.getAccess();
        if (!curPermissions) {
            return undefined;
        }
        const permissionActions = map(curPermissions, (permission) => permission && permission.actions);
        return hasPermission(permissionActions);
    }
}
