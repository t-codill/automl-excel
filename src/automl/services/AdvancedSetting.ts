import { AzureMachineLearningJasmineManagementModels } from "@vienna/jasmine";
import { ISettingsStepParams } from "../startRun/settingsSteps/ISettingsStepParams";
import { MetricType } from "./constants/MetricType";
import { RunType } from "./constants/RunType";

export class AdvancedSetting {
    public jobType: RunType;
    public column: string;
    public metric: AzureMachineLearningJasmineManagementModels.PrimaryMetric;
    public experimentTimeoutMinutes: number;
    public maxIteration: number;
    public experimentExitScore: number | null;
    public maxConcurrent: number;
    public maxCores: number;
    public preprocess: boolean;
    public nCrossValidations: number;
    public blacklistAlgos: string[] | null;
    public validationSize: number | null;
    public timeSeriesColumn: string | null;
    public maxHorizon: number | null;
    public grainColumns: string[] | null;

    // tslint:disable-next-line:cyclomatic-complexity
    constructor(input: ISettingsStepParams) {
        this.jobType = input.jobType || "classification";
        this.column = input.column || "";
        this.metric = input.metric || MetricType[this.jobType][0];
        this.experimentTimeoutMinutes = Number(input.trainingJobTime) || 60;
        this.maxIteration = Number(input.maxIteration) || 100;
        this.experimentExitScore = Number(input.metricThreshold) || null;
        this.maxConcurrent = Number(input.maxConcurrentIterations) || 1;
        this.maxCores = Number(input.maxCores) || -1;
        this.preprocess = input.preprocessing === undefined ? true : input.preprocessing;
        this.blacklistAlgos = input.blacklistAlgos;
        this.nCrossValidations = Number(input.crossValidationNumber) || 5;
        this.validationSize = Number(input.percentageValidation) / 100 || null;
        this.timeSeriesColumn = input.timeSeriesColumn || null;
        this.maxHorizon = Number(input.maxHorizon) || null;
        this.grainColumns = input.grainColumns;
    }
}
