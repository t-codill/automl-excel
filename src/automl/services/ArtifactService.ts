import { ArtifactAPI, ArtifactAPIModels } from "@vienna/artifact";
import { blob2string } from "../common/utils/blob2string";
import { IServiceBaseProps } from "./ServiceBase";
import { ServiceBaseNonArm } from "./ServiceBaseNonArm";

const artifactClientApiVersion = "1.0";
const runOrigin = "ExperimentRun";
const modelPath = "outputs/model.pkl";

export interface IArtifact {
    origin: string;
    container: string;
    path: string;
}

export interface IRunDataForArtifact {
    runId: string | undefined;
    parentRunId: string | undefined;
    status: string | undefined;
}

async function getArtifactContent(artifact: ArtifactAPIModels.GetArtifactContentByIdResponse): Promise<string | undefined> {
    const blobBody = await artifact.blobBody;
    if (blobBody) {
        return blob2string(blobBody);
    }
    return undefined;
}
export class ArtifactService extends ServiceBaseNonArm<ArtifactAPI> {
    constructor(props: IServiceBaseProps) {
        super(props, ArtifactAPI, props.discoverUrls.history);
    }
    public async tryGetContentForRun(runId: string, path: string): Promise<string | null | undefined> {
        return this.tryGetContent({ origin: runOrigin, container: runId, path });
    }

    public async tryGetContent(artifact: IArtifact): Promise<string | null | undefined> {
        return this.trySend(async (client, abortSignal) => {
            const artifactContent = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    path: artifact.path,
                    abortSignal
                });
            return getArtifactContent(artifactContent);
        });
    }

    public async getContent(artifact: IArtifact): Promise<string | undefined> {
        return this.send(async (client, abortSignal) => {
            const artifactContent = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    path: artifact.path,
                    abortSignal
                });
            return getArtifactContent(artifactContent);
        });
    }

    public async getAllContents(params: IArtifact[]): Promise<Array<string | undefined> | undefined> {
        return this.parallelGetAllValues(params, async (param, client, abortSignal) => {
            const artifact = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                param.origin,
                param.container,
                artifactClientApiVersion, {
                    path: param.path,
                    abortSignal
                });
            return getArtifactContent(artifact);
        });
    }

    public async getAllArtifactsForRuns(runs: string[]): Promise<ArtifactAPIModels.ArtifactDto[][] | undefined> {
        return this.parallelGetAllValuesWithContinuationToken(runs, async (runId, client, abortSignal, continuationToken) => {
            return client.getArtifactsInContainerOrPath2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                runOrigin,
                runId,
                artifactClientApiVersion, {
                    abortSignal,
                    continuationToken
                });
        });
    }

    public async getAllArtifacts(runId: string): Promise<ArtifactAPIModels.ArtifactDto[] | undefined> {
        return this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationToken) => {
            return client.getArtifactsInContainerOrPath2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                runOrigin,
                runId,
                artifactClientApiVersion, {
                    abortSignal,
                    continuationToken
                });
        });
    }

    public async tryGetArtifactUrl(artifact: IArtifact): Promise<ArtifactAPIModels.GetArtifactContentInformation2Response | null | undefined> {
        return this.trySend(async (client, abortSignal) => {
            return client.getArtifactContentInformation2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    abortSignal,
                    path: artifact.path
                });
        });
    }

    public async getModelUrl(run: IRunDataForArtifact): Promise<string | null | undefined> {
        if (!run || !run.runId || !run.parentRunId || run.status !== "Completed") {
            return null;
        }
        const result = await this.tryGetArtifactUrl({ container: run.runId, origin: runOrigin, path: modelPath });
        if (result === undefined) {
            return undefined;
        }
        if (!result || !result.contentUri) {
            return null;
        }
        return result.contentUri;
    }
}
