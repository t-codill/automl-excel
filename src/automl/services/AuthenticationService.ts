
const authenticationService = {
    async getToken(): Promise<string | undefined> {
        const response = await fetch("http://localhost:3001");
        const json = await response.json();
        return json.accessToken;
    }
};

export { authenticationService as AuthenticationService };
