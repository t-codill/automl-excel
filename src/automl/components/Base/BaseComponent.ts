import { mapValues } from "lodash";
import { BasicTypes } from "../../common/BasicTypes";
import { IDictionary } from "../../common/IDictionary";
import { IServiceBaseProps } from "../../services/ServiceBase";
import { BaseComponentContext } from "./BaseComponentContext";

export abstract class BaseComponent<
    TProp,
    TState,
    TServices extends {
        [key: string]: {
            dispose(): void;
            reset(): void;
        };
    }> extends BaseComponentContext<TProp, TState> {

    protected get autoRefreshEnabled(): boolean | undefined {
        return this._autoRefreshEnabled;
    }

    protected set autoRefreshEnabled(flag: boolean | undefined) {
        this._autoRefreshEnabled = flag;
    }

    protected services!: { [key in keyof TServices]: TServices[key] };

    protected abstract readonly serviceConstructors: { [key in keyof TServices]: (new (context: IServiceBaseProps, props: TProp) => TServices[key]) };

    protected abstract readonly getData?: () => Promise<void>;

    protected autoRefreshInMs: number | undefined;

    protected refreshTimeOut: number | undefined;

    private mounted = false;
    private _autoRefreshEnabled: boolean | undefined;

    constructor(props: TProp) {
        super(props);
        this.setState = this.setState.bind(this);
        this.refresh = this.refresh.bind(this);
        this.changeAutoRefresh = this.changeAutoRefresh.bind(this);
    }

    public async componentDidMount(): Promise<void> {
        this.mounted = true;
        this.services = mapValues(this.serviceConstructors, (cons) => {
            return new cons(this.context, this.props);
        }) as { [key in keyof TServices]: TServices[key] };
        await this.refresh();
    }

    public componentWillUnmount(): void {
        this.mounted = false;
        if (this.refreshTimeOut) {
            clearTimeout(this.refreshTimeOut);
        }
        Object.keys(this.services)
            .forEach((key) => {
                this.services[key].dispose();
            });
    }

    public setState<K extends keyof TState>(
        state: ((prevState: Readonly<TState>, props: Readonly<TProp>) => (Pick<TState, K> | TState | null)) | (Pick<TState, K> | TState | null),
        callback?: () => void
    ): void {
        if (this.mounted) {
            super.setState(state, callback);
        }
    }

    protected logUserAction(component: string, properties?: IDictionary<BasicTypes>): void {
        this.context.logger.logCustomEvent(
            `${this.context.pageName}_${component}_UserAction`,
            this.context,
            {
                ...properties,
                pageName: this.context.pageName,
                component
            }
        );
    }

    protected async refresh(): Promise<void> {
        if (!this.mounted) {
            return;
        }
        if (this.getData) {
            if (this.refreshTimeOut) {
                clearTimeout(this.refreshTimeOut);
            }
            this.resetService();
            this.context.setLoading(true);
            await this.getData();
            this.context.setLoading(false);
            if (this.mounted && this._autoRefreshEnabled && this.autoRefreshInMs) {
                this.refreshTimeOut = window.setTimeout(this.refresh.bind(this), this.autoRefreshInMs);
            }
        }
    }

    protected changeAutoRefresh(enable: boolean): void {
        if (!this.autoRefreshInMs) {
            return;
        }
        if (this.refreshTimeOut) {
            clearTimeout(this.refreshTimeOut);
        }
        if (!this.mounted) {
            return;
        }
        if (enable) {
            this.refreshTimeOut = window.setTimeout(this.refresh.bind(this), this.autoRefreshInMs);
        }
        this._autoRefreshEnabled = enable;
    }

    private resetService(): void {
        Object.keys(this.services)
            .forEach((key) => {
                this.services[key].reset();
            });
    }
}
