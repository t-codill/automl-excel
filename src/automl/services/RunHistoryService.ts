import { RunHistoryAPIs, RunHistoryAPIsModels } from "@vienna/runhistory";

import { flatMap, mergeWith, reduce, sortBy } from "lodash";
import { BasicTypes } from "../common/BasicTypes";
import { IDictionary } from "../common/IDictionary";
import { isRunCompleted } from "../common/utils/run";
import { safeParseJson } from "../common/utils/safeParseJson";
import { IArtifact } from "./ArtifactService";
import { IServiceBaseProps } from "./ServiceBase";
import { ServiceBaseNonArm } from "./ServiceBaseNonArm";

const artifactLocationRegV1 = /^aml:\/\/artifactId\/([^/]+)\/([^/]+)\/outputs\/(.*)$/;
const artifactLocationRegV2 = /^aml:\/\/artifactId\/([^/]+)\/([^/]+)\/(.*)$/;
export type IRunDtoWithExperimentName = RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto & { experimentName: string };
export type RunMetricType = BasicTypes | BasicTypes[] | IDictionary<BasicTypes[]>;

function getMetricValue(metric: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto): RunMetricType {
    if (!metric.cells
        || !metric.schema
        || !metric.schema.properties
        || !metric.schema.properties[0]
        || !metric.schema.properties[0].name) {
        return undefined;
    }
    if (metric.schema.properties.length === 1) {
        const propertyName = metric.schema.properties[0].name;
        // scalar
        if (metric.cells.length === 1) {
            return metric.cells[0][propertyName];
        }
        // array
        else {
            return metric.cells.map((c) => c[propertyName]);
        }
    }
    else {
        // create value from property: {property1: [], property2:[]}
        const tableValue: IDictionary<BasicTypes[]> = reduce<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsMetricSchemaPropertyDto, {}>(
            metric.schema.properties, (result, value) => {
                if (value.name) {
                    result[value.name] = [];
                }
                return result;
            }, {});
        return reduce<RunMetricType, {}>(metric.cells, (result, value) => {
            // push cells value to result
            // cell: {property1: 13}, result: {property1: [1], property2:[5]}
            // return {property1: [1,13], property2:[5]}
            mergeWith(result, value, (r, v) => {
                return [...r, v];
            });
            return result;
        }, tableValue);
    }
}
export class RunHistoryService extends ServiceBaseNonArm<RunHistoryAPIs> {
    constructor(props: IServiceBaseProps) {
        super(props, RunHistoryAPIs, props.discoverUrls.history);
    }
    public getArtifactBackedMetricNames(metrics: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto[]): Array<[string, IArtifact]> {
        return reduce(metrics, (result, m) => {
            if (m.metricType) {
                let artifactLocationReg: RegExp;
                switch (m.metricType) {
                    case "azureml.v1.residuals":
                    case "azureml.v1.predictions":
                    case "azureml.v1.confusion_matrix":
                    case "azureml.v1.accuracy_table":
                        artifactLocationReg = artifactLocationRegV1;
                        break;
                    case "azureml.v2.residuals":
                    case "azureml.v2.predictions":
                    case "azureml.v2.confusion_matrix":
                    case "azureml.v2.accuracy_table":
                        artifactLocationReg = artifactLocationRegV2;
                        break;
                    default:
                        return result;
                }
                const match = artifactLocationReg.exec(m.dataLocation || "");
                if (match && m.name) {
                    result.push(
                        [
                            m.name,
                            {
                                origin: match[1],
                                container: match[2],
                                path: match[3]
                            }
                        ]);
                }
            }
            return result;
        }, [] as Array<[string, IArtifact]>);
    }
    public mergeMetrics(
        metrics: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto[],
        artifactBackedMetricNames: string[],
        artifacts: Array<string | undefined>): IDictionary<RunMetricType> {
        return reduce<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto, IDictionary<RunMetricType>>(metrics, (result, m) => {
            if (m.name) {
                const index = artifactBackedMetricNames.indexOf(m.name);
                result[m.name] = (index >= 0 && artifacts[index])
                    ? safeParseJson(artifacts[index])
                    : getMetricValue(m);
            }
            return result;
        }, {});

    }

    public async getExperiment(experimentName: string): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto | undefined> {

        return this.send(async (client, abortSignal) => {
            return client.getExperiment(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experimentName, {
                abortSignal
            });
        });
    }

    public async getRun(runId: string, experimentName: string): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined> {
        return this.send(async (client, abortSignal) => {
            return client.getRunDetails(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experimentName, runId, {
                abortSignal
            });
        });
    }

    public async getRunMetrics(runId: string, experimentName: string): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto[] | undefined> {
        return this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationtoken) => {
            return client.getRunMetrics(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experimentName, {
                abortSignal,
                continuationtoken,
                filter: `RunId eq '${runId}'`
            });
        });
    }

    public async getChildRuns(runId: string, experimentName: string): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto[] | undefined> {
        const runs = await this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationtoken) => {
            return client.getChildRuns(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experimentName, runId,
                {
                    abortSignal,
                    continuationtoken
                });
        });
        if (!runs) {
            return undefined;
        }
        return sortBy(runs, (run) => run.properties ? parseInt(run.properties.iteration, 10) : undefined);
    }

    public async getChildRunMetrics(
        childRuns: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto[],
        metricsNames: string[],
        experimentName: string): Promise<Array<IDictionary<RunMetricType>> | undefined> {

        if (childRuns.length === 0) {
            return [];
        }

        const completedRuns = childRuns
            .filter(isRunCompleted);
        if (completedRuns.length < 1) {
            return childRuns.map(() => this.mergeMetrics([], [], []));
        }

        const metrics = await this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationToken) => {
            return client.queryRunMetrics(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experimentName, {
                abortSignal,
                queryParams: {
                    filter: `(${
                        completedRuns
                            .map((run) => `RunId eq '${run.runId}'`)
                            .join(" or ")
                        }) and (${
                        metricsNames.map((name) => `name eq '${name}'`)
                            .join(" or ")
                        })`,
                    continuationToken,
                    top: 1000
                }
            });
        });
        if (!metrics) { return undefined; }

        return childRuns.map((run) => this.mergeMetrics(metrics.filter((m) => m.runId === run.runId), [], []));
    }

    public async listExperiments(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto[] | undefined> {
        return this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationtoken) => {
            return client.getExperiments(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, {
                abortSignal,
                continuationtoken
            });
        });
    }

    // TODO: This is a temporary function only used by UX-Runner
    // This will be removed after we expose ODATA filter for getRunList
    public async getRuns(experiment: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto): Promise<IRunDtoWithExperimentName[] | undefined> {
        const allRuns = await this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationtoken) => {
            return client.getRuns(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, experiment.name || "",
                {
                    abortSignal,
                    continuationtoken,
                    filter: `Hidden eq false and parentRunId eq null and runType eq "automl"`,
                    top: 1000
                });
        });
        if (!allRuns) {
            return undefined;
        }

        return allRuns.map(
            ((run) => (
                {
                    ...run,
                    experimentName: experiment.name || ""
                })));
    }

    public async getRunList(): Promise<IRunDtoWithExperimentName[] | undefined> {
        const allExperiments = await this.listExperiments();
        if (!allExperiments) {
            return undefined;
        }
        const allRuns = await this.parallelGetAllValuesWithContinuationToken(allExperiments, async (exp, client, abortSignal, continuationtoken) => {
            return client.getRuns(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName, exp.name || "",
                {
                    abortSignal,
                    continuationtoken,
                    filter: `Hidden eq false and parentRunId eq null and runType eq "automl"`,
                    top: 1000
                });
        });
        if (!allRuns) {
            return undefined;
        }

        return flatMap(
            allRuns.map(
                (com, idx) =>
                    com.map((run) => (
                        {
                            ...run,
                            experimentName: allExperiments[idx].name || ""
                        }))));
    }

    public async updateTag(
        run: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto,
        experimentName: string,
        name: string,
        value: string
    ): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined> {
        return this.send(async (client, abortSignal) => {
            if (!run.runId) {
                return undefined;
            }
            const tags = { ...run.tags };
            tags[name] = value;
            return client.addOrModifyRun(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                experimentName,
                run.runId,
                {
                    createRunDto: {
                        runId: run.runId,
                        tags
                    },
                    abortSignal
                });
        });
    }

    public async updateTags(
        run: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto,
        experimentName: string,
        names: string[],
        values: string[]
    ): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined> {
        if (!names || !values || names.length !== values.length) {
            return undefined;
        }
        return this.send(async (client, abortSignal) => {
            if (!run.runId) {
                return undefined;
            }
            const tags = { ...run.tags };
            for (let i = 0; i < names.length; i++) {
                tags[names[i]] = values[i];
            }
            return client.addOrModifyRun(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                experimentName,
                run.runId,
                {
                    createRunDto: {
                        runId: run.runId,
                        tags
                    },
                    abortSignal
                });
        });
    }
}
