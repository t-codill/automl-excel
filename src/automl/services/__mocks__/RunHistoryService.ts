import { RunHistoryAPIsModels } from "@vienna/runhistory";
import { classificationSuccessRun } from "../../childRun/__data__/classificationSuccessRun";
import { BasicTypes } from "../../common/BasicTypes";
import { IDictionary } from "../../common/IDictionary";
import { parentSuccessRun } from "../../parentRun/__data__/parentSuccessRun";
import { IArtifact } from "../ArtifactService";

export type IRunDtoWithExperimentName = RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto & { experimentName: string };
export type RunMetricType = BasicTypes | BasicTypes[] | IDictionary<BasicTypes[]>;

export class RunHistoryService {
    public getArtifactBackedMetricNames(): Array<[string, IArtifact]> {
        const artifact: IArtifact = {
            origin: "",
            container: "",
            path: ""
        };
        return [["artifact", artifact]];
    }
    public mergeMetrics(): IDictionary<RunMetricType> {
        if (classificationSuccessRun.runMetrics) {
            return classificationSuccessRun.runMetrics;
        }
        return {};
    }

    public async getExperiment(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto | undefined> {
        return undefined;
    }

    public async getRun(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined> {
        return classificationSuccessRun.run;
    }

    public async getRunMetrics(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunMetricDto[] | undefined> {
        return [classificationSuccessRun.runMetrics];
    }

    public async getChildRuns(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto[] | undefined> {
        return undefined;
    }

    public async getChildRunMetrics(): Promise<Array<IDictionary<RunMetricType>> | undefined | undefined> {
        return undefined;
    }

    public async listExperiments(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsExperimentDto[] | undefined> {
        return [
            {
                createdUtc: new Date(2019, 1, 1, 1, 1, 1, 1),
                description: "Description 1",
                experimentId: "Experiment Id 1",
                name: "Experiment Name 1"
            },
            {
                createdUtc: new Date(2019, 2, 2, 2, 2, 2, 2),
                description: "Description 2",
                experimentId: "Experiment Id 2",
                name: "Experiment Name 2"
            },
            {}
        ];
    }

    public async getRunList(): Promise<IRunDtoWithExperimentName[] | undefined> {
        return [
            {
                experimentName: "experiment1",
                runId: "run1"
            },
            {
                experimentName: "experiment2",
                runId: "run2"
            }
        ];
    }

    public async updateTag(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined> {
        return parentSuccessRun.run;
    }

    public async updateTags(): Promise<RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined> {
        return parentSuccessRun.run;
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
