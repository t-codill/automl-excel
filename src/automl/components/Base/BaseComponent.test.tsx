import { shallow, ShallowWrapper } from "enzyme";
import * as React from "react";
import { RunHistoryService } from "../../services/RunHistoryService";
import { BaseComponent } from "./BaseComponent";

jest.unmock("./BaseComponent");
jest.mock("../../services/RunHistoryService");

interface IBaseComponentTesterProps {
    autoRefreshInMs?: number;
    getData?(): Promise<void>;
}
class BaseComponentTester extends BaseComponent<IBaseComponentTesterProps, { testState: boolean }, { history: RunHistoryService }> {
    protected getData?: () => Promise<void>;
    protected serviceConstructors = { history: RunHistoryService };
    constructor(props: IBaseComponentTesterProps) {
        super(props);
        this.autoRefreshInMs = props.autoRefreshInMs;
        this.getData = props.getData;
    }
    public render(): React.ReactNode {
        return <div />;
    }

    public changeAutoRefresh(enable: boolean): void {
        super.changeAutoRefresh(enable);
    }

    public getAutoRefreshEnabled(): boolean | undefined {
        return this.autoRefreshEnabled;
    }

    public getServices(): { history: RunHistoryService } {
        return this.services;
    }
}

describe("BaseComponent", () => {
    it("render", () => {
        const tree = shallow(<BaseComponentTester />);
        expect(tree)
            .toMatchSnapshot();
    });

    describe("refresh", () => {

        let setTimeoutSpy: jest.SpyInstance<ReturnType<Window["setTimeout"]>>;
        let clearTimeoutSpy: jest.SpyInstance<ReturnType<Window["clearTimeout"]>>;
        let refreshFunction: (() => Promise<void>) | undefined;
        beforeEach(() => {
            refreshFunction = undefined;
            setTimeoutSpy = jest.spyOn(window, "setTimeout");
            setTimeoutSpy.mockImplementation((func: () => Promise<void>) => {
                refreshFunction = func;
                return 1;
            });
            clearTimeoutSpy = jest.spyOn(window, "clearTimeout");
            clearTimeoutSpy.mockImplementation(() => {
                return;
            });
        });
        it("should not change auto refresh is autoRefreshInMs is not set", () => {
            const tree = shallow<BaseComponentTester>(<BaseComponentTester />);
            tree.instance()
                .changeAutoRefresh(true);
            expect(setTimeoutSpy)
                .toBeCalledTimes(0);
        });
        it("should set auto refresh", () => {
            const tree = shallow<BaseComponentTester>(<BaseComponentTester autoRefreshInMs={30000} />);
            tree.instance()
                .changeAutoRefresh(true);
            expect(setTimeoutSpy)
                .toBeCalledTimes(1);
        });
        it("should not set auto refresh is unmounted", () => {
            const tree = shallow<BaseComponentTester>(<BaseComponentTester autoRefreshInMs={30000} />);
            tree.instance()
                .componentWillUnmount();
            tree.instance()
                .changeAutoRefresh(true);
            expect(setTimeoutSpy)
                .toBeCalledTimes(0);
        });
        it("should disable auto refresh", () => {
            const tree = shallow<BaseComponentTester>(<BaseComponentTester autoRefreshInMs={30000} />);
            tree.instance()
                .changeAutoRefresh(true);
            setTimeoutSpy.mockClear();
            clearTimeoutSpy.mockClear();
            tree.instance()
                .changeAutoRefresh(false);
            expect(setTimeoutSpy)
                .toBeCalledTimes(0);
            expect(clearTimeoutSpy)
                .toBeCalledTimes(1);
            expect(tree.instance()
                .getAutoRefreshEnabled())
                .toBe(false);
        });
        it("should clear timeout when unload", () => {
            const tree = shallow<BaseComponentTester>(<BaseComponentTester autoRefreshInMs={30000} />);
            tree.instance()
                .changeAutoRefresh(true);
            setTimeoutSpy.mockClear();
            clearTimeoutSpy.mockClear();
            tree.instance()
                .componentWillUnmount();
            expect(clearTimeoutSpy)
                .toBeCalledTimes(1);
        });

        describe("refresh", () => {
            it("should set loading and call get data in refresh", (done) => {
                const props = {
                    autoRefreshInMs: 30000,
                    async getData(): Promise<void> { return; }
                };
                const setLoadingSpy = jest.spyOn(BaseComponent.prototype.context, "setLoading");
                const getDataSpy = jest.spyOn(props, "getData");
                shallow<BaseComponentTester>(<BaseComponentTester {...props} />);
                setImmediate(() => {
                    expect(getDataSpy)
                        .toBeCalledTimes(1);
                    expect(setLoadingSpy)
                        .toBeCalledTimes(2);
                    done();
                });
            });
            it("should clear timeout and set new timeout if auto refresh is enabled", async (done) => {
                const props = {
                    autoRefreshInMs: 30000,
                    async getData(): Promise<void> { return; }
                };
                const tree = shallow<BaseComponentTester>(<BaseComponentTester {...props} />);
                tree.instance()
                    .changeAutoRefresh(true);
                setImmediate(async () => {
                    clearTimeoutSpy.mockClear();
                    setTimeoutSpy.mockClear();
                    if (refreshFunction) {
                        refreshFunction();
                    }
                    setImmediate(() => {
                        expect(clearTimeoutSpy)
                            .toBeCalledTimes(1);
                        expect(setTimeoutSpy)
                            .toBeCalledTimes(1);
                        done();
                    });
                });
            });
            it("unmount component should not refresh", (done) => {
                const props = {
                    autoRefreshInMs: 30000,
                    async getData(): Promise<void> { return; }
                };
                const tree = shallow<BaseComponentTester>(<BaseComponentTester {...props} />);
                tree.instance()
                    .changeAutoRefresh(true);
                tree.instance()
                    .componentWillUnmount();
                const getDataSpy = jest.spyOn(props, "getData");
                if (refreshFunction) {
                    refreshFunction();
                }
                setImmediate(() => {
                    expect(getDataSpy)
                        .toBeCalledTimes(0);
                    done();
                });
            });
        });
    });
    describe("services", () => {
        let runHistoryServiceDisposeSpy: jest.SpyInstance<ReturnType<RunHistoryService["dispose"]>>;
        beforeEach(() => {
            runHistoryServiceDisposeSpy = jest.spyOn(RunHistoryService.prototype, "dispose");
        });
        it("should create service", () => {
            const tree = shallow<BaseComponentTester>(<BaseComponentTester />);
            expect(tree.instance()
                .getServices().history)
                .toBeDefined();
        });
        it("should dispose service", () => {
            const tree = shallow<BaseComponentTester>(<BaseComponentTester />);
            tree.instance()
                .componentWillUnmount();
            expect(runHistoryServiceDisposeSpy)
                .toBeCalledTimes(1);
        });
    });

    describe("setState", () => {
        let tree: ShallowWrapper<BaseComponentTester["props"], BaseComponentTester["state"], BaseComponentTester>;
        beforeAll(() => {
            tree = shallow<BaseComponentTester>(<BaseComponentTester />);
        });
        it("init state is undefined", () => {
            expect(tree.state())
                .toBeNull();
        });
        it("should setState", () => {
            tree.instance()
                .setState({ testState: true });
            expect(tree.state())
                .toEqual({ testState: true });
        });
        it("should not setState after unload", () => {
            tree.instance()
                .componentWillUnmount();
            tree.instance()
                .setState({ testState: false });
            expect(tree.state())
                .toEqual({ testState: true });
        });
    });
});
