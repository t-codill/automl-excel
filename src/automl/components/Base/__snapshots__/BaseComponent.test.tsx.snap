// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseComponent render 1`] = `<div />`;
